src/serac/physics/thermal_mechanics_functional.hpp

```diff
@@ -0,0 +1,350 @@
+
+    state_.push_back(thermal_functional_.temperature());
+    state_.push_back(solid_functional_.velocity());
+    state_.push_back(solid_functional_.displacement());
+
+    thermal_functional_.setParameter(solid_functional_.displacement(), 0);
+    solid_functional_.setParameter(thermal_functional_.temperature(), 0);
+
+    coupling_ = serac::CouplingScheme::OperatorSplit;
+  }
+
+  /**
+   * @brief Complete the initialization and allocation of the data structures.
+   *
+   * @note This must be called before AdvanceTimestep().
+   */
+  void completeSetup() override
+  {
+    SLIC_ERROR_ROOT_IF(coupling_ != serac::CouplingScheme::OperatorSplit,
+                       "Only operator split is currently implemented in the thermal structural solver.");
+
+    thermal_functional_.completeSetup();
+    solid_functional_.completeSetup();
+  }
+
+  /**
+   * @brief register the provided FiniteElementState object as the source of values for parameter `i`
+   *
+   * @param parameter_state the values to use for the specified parameter
+   * @param i the index of the parameter
+   */
+  void setParameter(const FiniteElementState& parameter_state, size_t i)
+  {
+    thermal_functional_.setParameter(parameter_state, i + 1);  // offset for displacement field
+    solid_functional_.setParameter(parameter_state, i + 1);    // offset for temperature field
+  }
+
+  /**
+   * @brief Advance the timestep
+   *
+   * @param[inout] dt The timestep to attempt. This will return the actual timestep for adaptive timestepping
+   * schemes
+   * @pre completeSetup() must be called prior to this call
+   */
+  void advanceTimestep(double& dt) override
+  {
+    if (coupling_ == serac::CouplingScheme::OperatorSplit) {
+      double initial_dt = dt;
+      thermal_functional_.advanceTimestep(dt);
+      solid_functional_.advanceTimestep(dt);
+      SLIC_ERROR_ROOT_IF(std::abs(dt - initial_dt) > 1.0e-6,
+                         "Operator split coupled solvers cannot adaptively change the timestep");
+    } else {
+      SLIC_ERROR_ROOT("Only operator split coupling is currently implemented");
+    }
+
+    cycle_ += 1;
+  }
+
+  /**
+   * @brief Create a shared ptr to a quadrature data buffer for the given material type
+   *
+   * @tparam T the type to be created at each quadrature point
+   * @param initial_state the value to be broadcast to each quadrature point
+   * @return std::shared_ptr< QuadratureData<T> >
+   */
+  template <typename T>
+  std::shared_ptr<QuadratureData<T>> createQuadratureDataBuffer(T initial_state)
+  {
+    return solid_functional_.createQuadratureDataBuffer(initial_state);
+  }
+
+  /**
+   * @brief This is an adaptor class that makes a thermomechanical material usable by
+   * the thermal module, by discarding the solid-mechanics-specific information
+   *
+   * @tparam ThermalMechanicalMaterial the material model being wrapped
+   */
+  template <typename ThermalMechanicalMaterial>
+  struct ThermalMaterialInterface {
```

Comment:
Should this be in the materials directory? I can see it going either here or there.

Reply:
It's glue code between the materials and the physics module, which does make it tricky to assign a home. I was also conflicted, but I prefer the current arrangement because the glue code is a consequence of the operator split solver algorithm. Putting it with the materials will make material model developers confront the solver algorithm details. I'd prefer to let them focus on the physics and ignore how the solver will use their model.